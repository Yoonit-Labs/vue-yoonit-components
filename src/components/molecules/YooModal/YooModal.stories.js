import yooModal from './YooModal'

export default {
  title: 'Yoonit/Modal',
  component: yooModal
}

const Template = (args, { argTypes }) => ({
  props: Object.keys(argTypes),
  components: { yooModal },
  template:
    `<yoo-modal v-bind="$props">
        <template v-slot:modal>
          <h3>MODAL</h3>
          <p>Conte√∫do do Modal</p>
        </template>
    </yoo-modal>`
})

export const Custom = Template.bind({})
