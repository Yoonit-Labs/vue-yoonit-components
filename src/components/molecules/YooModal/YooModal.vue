<template lang="pug">
  .backdrop(
    @click.self.stop="doCloseModal"
  )
    YooFlexLayout.yoo-modal(
      flexDirection="column"
      flexWrap="wrap"
      justifyContent="center"
      alignItems="center"
    )
      slot(
        name="modal"
      )
</template>

<script>

import YooFlexLayout from '@/components/quarks/YooFlexLayout/YooFlexLayout'

export default {
  name: 'YooModal',
  components: {
    YooFlexLayout
  },
  data: () => ({}),
  beforeMount () {
    window.addEventListener('keyup', this.onEscapeKeyUp)
  },
  beforeDestroy () {
    window.removeEventListener('keyup', this.onEscapeKeyUp)
  },
  computed: {},
  methods: {
    doCloseModal (event) {
      this.$emit('onClose', event)
    },
    onEscapeKeyUp (event) {
      if (event.which === 27) {
        this.doCloseModal(event)
      }
    }
  }
}
</script>

<style src="./YooModal.sass" lang="sass" scoped></style>
