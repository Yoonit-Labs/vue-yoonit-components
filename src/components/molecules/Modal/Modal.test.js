/**
 * ██╗   ██╗ ██████╗  ██████╗ ███╗   ██╗██╗████████╗
 * ╚██╗ ██╔╝██╔═══██╗██╔═══██╗████╗  ██║██║╚══██╔══╝
 *  ╚████╔╝ ██║   ██║██║   ██║██╔██╗ ██║██║   ██║
 *   ╚██╔╝  ██║   ██║██║   ██║██║╚██╗██║██║   ██║
 *    ██║   ╚██████╔╝╚██████╔╝██║ ╚████║██║   ██║
 *    ╚═╝    ╚═════╝  ╚═════╝ ╚═╝  ╚═══╝╚═╝   ╚═╝
 *
 * https://yoonit.dev - about@yoonit.dev
 *
 * Vue Yoonit Components
 * VueJS Atomic Design System framework
 *
 * Vitória Costa, Tiago Brito, Fernando Junior, Sabrina Sampaio, Gabriel Mule, Gabriel Moraes, Gabriel Rizzo & Luigui Delyer @ 2020-2021
 */

import { shallowMount } from '@vue/test-utils'
import YooModal from '@/components/molecules/Modal/Modal.vue'
import PropsConfig from '@/components/molecules/Modal/Modal.config'

const classBlock = 'yoo-modal'

const mountYooModal = () => {
  return shallowMount(YooModal)
}

describe('YooModal Component', () => {
  let wrapper
  beforeEach(() => {
    wrapper = mountYooModal()
  })

  it('Matches Snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('Loads the component HTML', () => {
    expect(wrapper.find('.backdrop').exists()).toBe(true)
    expect(wrapper.find('.yoo-modal').exists()).toBe(true)
  })

  describe('Props', () => {
    describe('fill', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.fill.options.includes(YooModal.props.fill.default)).toBe(true)
      })
      PropsConfig.fill.options.forEach(fill => {
        it(`Includes fill class: .${classBlock}__fill--${fill}`, async () => {
          await wrapper.setProps({ fill })
          expect(wrapper.find(`.${classBlock}__fill--${fill}`).exists()).toBe(true)
        })
      })
    })

    describe('borderStyle', () => {
      it('Has a valid default value', () => {
        expect(PropsConfig.borderStyle.options.includes(YooModal.props.borderStyle.default)).toBe(true)
      })
      PropsConfig.borderStyle.options.forEach(borderStyle => {
        it(`Includes borderStyle class: .${classBlock}__border--${borderStyle}`, async () => {
          await wrapper.setProps({ borderStyle })
          expect(wrapper.find(`.${classBlock}__border--${borderStyle}`).exists()).toBe(true)
        })
      })
    })
  })

  describe('Events', () => {
    describe('Close Modal', () => {
      it('Emits Click Event', async () => {
        await wrapper.trigger('click')
        expect(wrapper.emitted()).toHaveProperty('onClose')
      })
      it('doCloseModal function emits onClose', async () => {
        await wrapper.vm.doCloseModal()
        expect(wrapper.emitted()).toHaveProperty('onClose')
      })
      it('Click runs doCloseModal method', async () => {
        const logger = jest.spyOn(wrapper.vm, 'doCloseModal')
        await wrapper.trigger('click')
        expect(logger).toHaveBeenCalled()
      })
      it('Emits onClose on ESC keyup', async () => {
        var event = new KeyboardEvent('keyup', { which: 27 })
        window.dispatchEvent(event)
        expect(wrapper.emitted('onClose')).toBeTruthy()
      })
      it('Does not emit on other key codes', async () => {
        var event = new KeyboardEvent('keyup', { which: 13 })
        window.dispatchEvent(event)
        expect(wrapper.emitted('onClose')).toBeFalsy()
      })
      it('Does not emit after destroy component', async () => {
        var event = new KeyboardEvent('keyup', { which: 27 })
        wrapper.destroy()
        window.dispatchEvent(event)
        expect(wrapper.emitted('onClose')).toBeFalsy()
      })
    })
  })
})
