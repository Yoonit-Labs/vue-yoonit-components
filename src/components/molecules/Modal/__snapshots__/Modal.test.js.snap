// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`YooModal Component Matches Snapshot 1`] = `
<div class="backdrop">
  <yooflexlayout-stub flexdirection="column" flexwrap="wrap" justifycontent="center" alignitems="center" aligncontent="stretch" gap="0" class="yoo-modal m__l--l m__r--l yoo-modal__fill--light yoo-modal__border--square"></yooflexlayout-stub>
</div>
`;
